src/app/background/sagas/refreshMatchingContexts.ts

```diff
@@ -36,29 +35,42 @@ const callAndRetry = createCallAndRetry({
   }
 });
 
-export function* refreshMatchingContextsSaga() {
-  const subscriptions = yield select(getSubscriptions);
-
+export function* fetchMatchingContextsForContributor(
+  contributorId: ContributorId
+) {
   const matchingContexts = yield callAndRetry(
     fetchMatchingContexts,
-    subscriptions
+    contributorId
   );
-
   if (matchingContexts) {
-    yield put(receivedMatchingContexts(matchingContexts));
+    yield put(receivedMatchingContexts(contributorId, matchingContexts));
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore To be typed properly
+export function* refreshMatchingContextsPeriodically() {
+  const subscriptions: ReturnType<typeof getSubscriptions> = yield select(
+    getSubscriptions
+  );
+
+  for (const subscription of subscriptions) {
+    yield fetchMatchingContextsForContributor(subscription);
   }
 
   if (refreshInterval > 0) {
     yield delay(refreshInterval);
-    yield put(refreshMatchingContexts());
+    yield refreshMatchingContextsPeriodically();
   }
 }
 
+function* fetchNewlySubscribedMatchingContexts(
+  subscribeAction: SubscribeAction
+) {
+  yield fetchMatchingContextsForContributor(subscribeAction.payload);
+}
+
 export default function* refreshMatchingContextsRootSaga() {
-  yield refreshMatchingContextsSaga();
-  yield debounce(
-    30_000,
-    [SUBSCRIBE, UNSUBSCRIBE, REFRESH_MATCHING_CONTEXTS],
-    refreshMatchingContextsSaga
-  );
+  yield fork(refreshMatchingContextsPeriodically);
```

Comment:
This is probably enough for our needs, but I probably would have kept an action triggering a refresh..
1. A command asking to refresh/fetch (this way the app know we are `fetching`)
2. Then dispatch another command for each contributor... and wait for all of them to comeback

A bit like the collection system implemented on the profile side.
But I know this is quite an heavy system and might be too much for our needs.

Reply:
I tend more and more to prefer having the control of the entire flow centralized in a saga …

Sure, we may need an action to let the UI know that we’re fetching … We don’t know what to do with this information right now, but the day we do it’s super easy to do : We just `put` a `REFRESHING_MATCHING_CONTEXT` **event** instead of `put`ing a **command** that we wait for in another subsaga …

It gives us even more control, with the ability to `put` intermediate actions, in order to notify about retries, easily parallelize requests (what I didn’t do yet), …

Also, doing this one, the whole fetching MC system is a bit more isolated and independent from the rest of the system